lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -800,6 +800,45 @@ static void scheduleDistanceOneDependencies(scf::ForOp forOp,
   }
 }
 
+static void scheduleRemainingToLastStage(scf::ForOp forOp,
+                                         CoarseSchedule &schedule,
+                                         CoarseSchedule::Cluster afterPrologue,
+                                         int numStages) {
+  // Assign the rest of the ops to the last stage.
+  // Take care of the ordering of the ops - uses cannot be scheduled to the
+  // cluster before the definition.
+  DenseMap<Operation *, CoarseSchedule::Cluster> opToCluster;
+  for (auto &op : forOp.getBody()->without_terminator()) {
+    if (schedule.count(&op) == 0) {
+      opToCluster[&op] = afterPrologue;
+    }
+  }
+  SmallVector<Operation *> queue;
+  for (auto [op, stage, cluster] : schedule.getOpsInOrder(forOp)) {
+    // We really only care about the producers from the last stage.
+    // Others will be scheduled before these ops anyway.
+    if (stage == numStages - 1) {
+      queue.push_back(op);
+    }
+  }
+  while (!queue.empty()) {
+    Operation *op = queue.pop_back_val();
+    for (auto user : op->getUsers()) {
+      if (opToCluster.count(user)) {
+        CoarseSchedule::Cluster userCluster = opToCluster[user];
+        CoarseSchedule::Cluster opCluster = schedule[op].second;
+        if (*userCluster < *opCluster) {
+          opToCluster[user] = opCluster;
+          queue.push_back(user);
+        }
+      }
+    }
+  }
```

Comment:
won't that re-order operations within the stage even if there is no `if`? I'm a bit worried this will re-order more than we want to. Ideally we keep the original order unless we have to change it to avoid generating poor schedule by accident.

Reply:
This should kick in only if the producer is being pushed to a later cluster, in which case it is functionally required that we push consumer down too. Consider this:
```
%cnd = ...
%1 = arith.select %cnd, ...
%2 = scf.if %cnd {
```
If `scf.if` will be pushed down, `%cnd` will be pushed down as well to keep the liverange short. But then we need to push select to the same cluster too. So it is not just about the downstream uses of `if`s that we need to push down.

Reply:
ah good, ok I missed that part.